README.md
# Tomcat Session Logout Listener

[![Release](https://img.shields.io/github/v/release/smarkwal/tomcat-session-logout-listener?label=Latest)](https://github.com/smarkwal/tomcat-session-logout-listener/releases/latest)
[![License](https://img.shields.io/github/license/smarkwal/tomcat-session-logout-listener?label=License)](https://github.com/smarkwal/tomcat-session-logout-listener/blob/main/LICENSE)
[![Issues](https://img.shields.io/github/issues/smarkwal/tomcat-session-logout-listener?label=Issues)](https://github.com/smarkwal/tomcat-session-logout-listener/issues)

[![Build](https://github.com/smarkwal/tomcat-session-logout-listener/actions/workflows/build.yml/badge.svg)](https://github.com/smarkwal/tomcat-session-logout-listener/actions/workflows/build.yml)
[![Tests](https://img.shields.io/sonar/tests/smarkwal_tomcat-session-logout-listener/main?label=Tests&server=https%3A%2F%2Fsonarcloud.io)](https://sonarcloud.io/component_measures?id=smarkwal_tomcat-session-logout-listener&metric=test_success_density&selected=smarkwal_tomcat-session-logout-listener%3Asrc%2Ftest%2Fjava%2Forg%2Fsession-logout-listener)
[![Coverage](https://img.shields.io/sonar/coverage/smarkwal_tomcat-session-logout-listener/main?label=Coverage&server=https%3A%2F%2Fsonarcloud.io)](https://sonarcloud.io/component_measures?id=smarkwal_tomcat-session-logout-listener&metric=coverage&view=list)
[![Quality](https://img.shields.io/sonar/quality_gate/smarkwal_tomcat-session-logout-listener/main?label=Quality&server=https%3A%2F%2Fsonarcloud.io)](https://sonarcloud.io/dashboard?id=smarkwal_tomcat-session-logout-listener)

Tomcat valve for remote session invalidation.

### Situation

A web application is hosted on [Apache Tomcat](https://tomcat.apache.org/).
Tomcat is configured to use an external user management system (e.g., LDAP) for authentication.
Upon successful authentication, Tomcat creates a session and stores the user's identity (principal) in that session.
The ID of the session is stored in a cookie on the user's browser.

For performance reasons, Tomcat usually does not re-validate the user's identity on every request.
This would require an expensive round-trip to the user management system.
Instead, Tomcat fully relies on the data in the session to authorize access to protected resources.

### Problem

If the user's password is changed, if the user is removed from some security roles, or if the user's account is disabled or deleted,
an already logged-in user can continue to work with the web application as if nothing has happened.
This can be a security problem.

It would be nice if the external user management system could notify Tomcat that the user's identity has changed.
Then, Tomcat could invalidate that user's session and force the user to log in again.

### Solution

This project provides a Tomcat valve that implements a web hook for remote session invalidation.
An external user management system (or an intermediate component) can call the web hook to invalidate a session.

This valve waits for HTTP requests sent to the endpoint `/session-logout-listener`.
When such a request is received, it is intercepted and processed by the valve.
The valve looks for a parameter with name `username`.
If the parameter is present, the valve iterates over all active Tomcat sessions and
invalidate all sessions with a principal name matching that parameter value.

```mermaid
sequenceDiagram
    participant Client
    participant Tomcat Valve
    participant Tomcat Session Manager
    Client->>+Tomcat Valve: Request to /session-logout-listener
    Tomcat Valve->>Tomcat Session Manager: Invalidate sessions
    Tomcat Valve->>-Client: "OK"
```

### Note

This project does not provide a solution for how to invoke the web hook.

Please consult the documentation of your user management system for how to be notified of changes to user identities.
Maybe it offers this functionality out of the box, maybe you have to implement some kind of plugin,
or maybe you have to write a custom component that periodically polls the user management system for changes.

## Requirements and compatibility

This valve has been tested with the following software components:

* Tomcat 7.0.109, Tomcat 8.5.84, and Tomcat 9.0.70.
* Java 8, Java 11, and Java 17.

Note: The valve is not compatible with Tomcat 10.0.0 or newer.

# Download

The latest release is available on [GitHub](https://github.com/smarkwal/tomcat-session-logout-listener/releases)
and [Maven Central](https://mvnrepository.com/artifact/net.markwalder/tomcat-session-logout-listener).

# Installation and configuration

Copy `tomcat-session-logout-listener-<version>.jar` to `$TOMCAT_HOME/lib/`.

Add the following line to `$TOMCAT_HOME/conf/server.xml` or `$TOMCAT_HOME/conf/context.xml`:

```xml
<Valve className="net.markwalder.tomcat.SessionLogoutListener"/>
```

The valve supports the following configuration attributes:

| Attribute  | Default value   | Description                                                                                                                                              |
|------------|-----------------|----------------------------------------------------------------------------------------------------------------------------------------------------------|
| `ipFilter` | `127.0.0.1,::1` | Comma-separated list of client IPv4 or IPv6 addresses and subnets that are allowed to call the web hook. The special value `*` matches all IP addresses. |
| `password` | (not set)       | Password required to call the web hook. Must be submitted as request parameter `password`.                                                              |

Example configuration:

```xml
<Valve className="net.markwalder.tomcat.SessionLogoutListener" 
       ipFilter="127.0.0.1,10.0.0.0/24,::1,fc00::/7" 
       password="my-secret-123!"
/>
```

Typical IP subnets for local and private address ranges:

* `127.0.0.0/8`
* `10.0.0.0/8`
* `172.16.0.0/12`
* `192.168.0.0/16`
* `fc00::/7`
* `fe80::/10`

### Logging

Add the following lines to `$TOMCAT_HOME/conf/logging.properties`:

```properties
# Enable debug logging for Tomcat Session Logout Listener
net.markwalder.tomcat.SessionLogoutListener.level = FINE
```

When a request is sent to the session logout endpoint, the following log messages are written:

```
FINE [http-nio-8080-exec-1] net.markwalder.tomcat.SessionLogoutListener.logoutUsers usernames: 'alice', 'bob'
FINE [http-nio-8080-exec-1] net.markwalder.tomcat.SessionLogoutListener.logoutUsers session: id='78F9279D...', principal='alice'
FINE [http-nio-8080-exec-1] net.markwalder.tomcat.SessionLogoutListener.logoutUsers session: id='1EF34298...', principal='bob'
```

For security reasons, only the first 8 characters of the session ID are logged.

# Integration

Send an HTTP request to the session logout endpoint with a parameter `username` set to the username of the user to be logged out.
You can use any HTTP client like `curl` or `wget`, or third-party HTTP or REST client libraries in your favorite programming language.

Example `GET` request:

```
GET /session-logout-listener?username=alice HTTP/1.1
```

Example `POST` request:

```
POST /session-logout-listener HTTP/1.1
Content-Type: application/x-www-form-urlencoded
Content-Length: 14

username=alice
```

This request will invalidate all sessions of user "alice".

It is possible to log out multiple users at once by specifying the parameter `username` multiple times.

Example `GET` request:

```
GET /session-logout-listener?username=alice&username=bob HTTP/1.1
```

Example `POST` request:

```
POST /session-logout-listener HTTP/1.1
Content-Type: application/x-www-form-urlencoded
Content-Length: 27

username=alice&username=bob
```

Note that if your web application is deployed in a context path other than `/`, you must include the context path in the request URL.

### Successful response

The session logout endpoint will return an HTTP response with status code 200 (OK), content type "text/plain", character set "UTF-8", and the text "OK" as body:

```
HTTP/1.1 200 OK
Content-Type: text/plain;charset=UTF-8
Content-Length: 2

OK
```

### Error response

If the client's IP address or password is not accepted by the valve, an HTTP response with status code 403 (Forbidden) and the text "Forbidden" is returned:

```
HTTP/1.1 403 Forbidden
Content-Type: text/plain;charset=UTF-8
Content-Length: 8

Forbidden
```

More details can be found in Tomcat's log file.

Example log messages:

```
WARN [http-nio-8080-exec-1] net.markwalder.tomcat.RemoteAddrCheck.test Remote address '123.45.67.89' does not match IP filter.
WARN [http-nio-8080-exec-1] net.markwalder.tomcat.PasswordCheck.test No password found in request.
WARN [http-nio-8080-exec-1] net.markwalder.tomcat.PasswordCheck.test Incorrect password.
```

### Clusters

If Tomcat is running in a cluster, the session logout endpoint must be called on all cluster nodes.

### Security considerations

The session logout endpoint can be protected by client IP address filtering and/or a password.
If you have additional security requirements, feel free to fork this project and implement your own security mechanism.

To prevent parameters from being logged as part of the URL, it is recommended to use `POST` requests instead of `GET` requests.

To prevent session IDs from leaking, only the first 8 characters of the session ID are logged.
But since these sessions have been invalidated, this should not be a security problem anyway.

# License

This project is licensed under the [MIT license](https://github.com/smarkwal/tomcat-session-logout-listener/blob/main/LICENSE).

# Build from source code

To build this project from source code, execute the following command:

```shell
# set JAVA_HOME to a JDK 11
export JAVA_HOME=/path/to/java-11

# build JAR file
./gradlew clean jar
```

The JAR file `tomcat-session-logout-listener-<version>.jar` can be found in `build/libs`.

### Run tests

Integration tests use [JUnit 5](https://junit.org/junit5/), [Testcontainers](https://www.testcontainers.org/), and [Docker](https://www.docker.com/) to run different Tomcat versions in containers.
Docker therefore must be installed on the build machine.

To run all tests, execute the following command:

```shell
./gradlew clean test
```

The test report can be found under `build/reports/tests/test/index.html`.


CHANGELOG.md
# Changelog

## 1.2.1-SNAPSHOT

* See [commits since version 1.2.0](https://github.com/smarkwal/tomcat-session-logout-listener/compare/v1.2.0...main).

## 1.2.0

* Feature: Support IPv6 in IP address filter.

## 1.1.0

* Feature: Support client IP address filter and password check.
* Bugfix: Skip sessions which have already expired.

## 1.0.0

* Initial release.

src/test/java/net/markwalder/tomcat/IpFilterTest.java
/*
 * MIT License
 *
 * Copyright (c) 2022 Stephan Markwalder
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package net.markwalder.tomcat;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import org.junit.jupiter.api.Test;

class IpFilterTest {

	private static final String PRIVATE_FILTER = "127.0.0.0/8,10.0.0.0/8,172.16.0.0/12,192.168.0.0/16";

	private static final String ALL_SUBNET = "0.0.0.0/0";
	private static final String LOCAL_SUBNET = "127.0.0.0/8";
	private static final String CLASS_A_SUBNET = "10.0.0.0/8";
	private static final String CLASS_B_SUBNET = "172.16.0.0/12";
	private static final String CLASS_C_SUBNET = "192.168.0.0/16";

	@Test
	void matches_localhost_range() {
		assertFalse(IpFilter.matches("126.255.255.255", PRIVATE_FILTER));
		assertTrue(IpFilter.matches("127.0.0.0", PRIVATE_FILTER));
		assertTrue(IpFilter.matches("127.0.0.1", PRIVATE_FILTER));
		assertTrue(IpFilter.matches("127.1.2.3", PRIVATE_FILTER));
		assertTrue(IpFilter.matches("127.255.255.255", PRIVATE_FILTER));
		assertFalse(IpFilter.matches("128.0.0.0", PRIVATE_FILTER));
		assertFalse(IpFilter.matches("123.45.67.89", PRIVATE_FILTER));
	}

	@Test
	void matches_class_A_private_range() {
		assertFalse(IpFilter.matches("9.255.255.255", PRIVATE_FILTER));
		assertTrue(IpFilter.matches("10.0.0.0", PRIVATE_FILTER));
		assertTrue(IpFilter.matches("10.1.2.3", PRIVATE_FILTER));
		assertTrue(IpFilter.matches("10.255.255.255", PRIVATE_FILTER));
		assertFalse(IpFilter.matches("11.0.0.0", PRIVATE_FILTER));
		assertFalse(IpFilter.matches("123.45.67.89", PRIVATE_FILTER));
	}

	@Test
	void matches_class_B_private_range() {
		assertFalse(IpFilter.matches("172.15.255.255", PRIVATE_FILTER));
		assertTrue(IpFilter.matches("172.16.0.0", PRIVATE_FILTER));
		assertTrue(IpFilter.matches("172.17.2.3", PRIVATE_FILTER));
		assertTrue(IpFilter.matches("172.31.255.255", PRIVATE_FILTER));
		assertFalse(IpFilter.matches("172.32.0.0", PRIVATE_FILTER));
		assertFalse(IpFilter.matches("123.45.67.89", PRIVATE_FILTER));
	}

	@Test
	void matches_class_C_private_range() {
		assertFalse(IpFilter.matches("192.167.255.255", PRIVATE_FILTER));
		assertTrue(IpFilter.matches("192.168.0.0", PRIVATE_FILTER));
		assertTrue(IpFilter.matches("192.168.1.2", PRIVATE_FILTER));
		assertTrue(IpFilter.matches("192.168.255.255", PRIVATE_FILTER));
		assertFalse(IpFilter.matches("192.169.0.0", PRIVATE_FILTER));
		assertFalse(IpFilter.matches("123.45.67.89", PRIVATE_FILTER));
	}

	@Test
	void matches_exact() {
		assertTrue(IpFilter.matches("123.45.67.89", "123.45.67.89"));
		assertFalse(IpFilter.matches("123.45.67.89", "1.2.3.4"));
	}

	@Test
	void matches_illegal_values() {
		assertFalse(IpFilter.matches(null, "10.0.0.0/8"));
		assertFalse(IpFilter.matches("10.0.0.1", null));
	}

	@Test
	void matchesAddress() {
		assertTrue(IpFilter.matchesAddress("127.0.0.1", "127.0.0.1"));
		assertFalse(IpFilter.matchesAddress("127.0.0.1", "10.0.0.1"));
	}

	@Test
	void matchesAddress_wildcard() {
		assertTrue(IpFilter.matchesAddress("127.0.0.1", "*"));
	}

	@Test
	void matchesAddress_illegal_values() {
		assertFalse(IpFilter.matchesAddress(null, "10.0.0.1"));
		assertFalse(IpFilter.matchesAddress("10.0.0.1", null));
		assertFalse(IpFilter.matchesAddress("10.0.0", "10.0.0.1"));
		assertFalse(IpFilter.matchesAddress("10.0.0.1", "10.0.0"));
	}

	@Test
	void matchesRange_local_subnet() {
		assertFalse(IpFilter.matchesRange("126.255.255.255", LOCAL_SUBNET));
		assertTrue(IpFilter.matchesRange("127.0.0.0", LOCAL_SUBNET));
		assertTrue(IpFilter.matchesRange("127.0.0.1", LOCAL_SUBNET));
		assertTrue(IpFilter.matchesRange("127.1.2.3", LOCAL_SUBNET));
		assertTrue(IpFilter.matchesRange("127.255.255.255", LOCAL_SUBNET));
		assertFalse(IpFilter.matchesRange("128.0.0.0", LOCAL_SUBNET));
		assertFalse(IpFilter.matchesRange("123.45.67.89", LOCAL_SUBNET));
	}

	@Test
	void matchesRange_class_A_subnet() {
		assertFalse(IpFilter.matchesRange("9.255.255.255", CLASS_A_SUBNET));
		assertTrue(IpFilter.matchesRange("10.0.0.0", CLASS_A_SUBNET));
		assertTrue(IpFilter.matchesRange("10.1.2.3", CLASS_A_SUBNET));
		assertTrue(IpFilter.matchesRange("10.255.255.255", CLASS_A_SUBNET));
		assertFalse(IpFilter.matchesRange("11.0.0.0", CLASS_A_SUBNET));
		assertFalse(IpFilter.matchesRange("123.45.67.89", CLASS_A_SUBNET));
	}

	@Test
	void matchesRange_class_B_subnet() {
		assertFalse(IpFilter.matchesRange("172.15.255.255", CLASS_B_SUBNET));
		assertTrue(IpFilter.matchesRange("172.16.0.0", CLASS_B_SUBNET));
		assertTrue(IpFilter.matchesRange("172.17.2.3", CLASS_B_SUBNET));
		assertTrue(IpFilter.matchesRange("172.31.255.255", CLASS_B_SUBNET));
		assertFalse(IpFilter.matchesRange("172.32.0.0", CLASS_B_SUBNET));
		assertFalse(IpFilter.matchesRange("123.45.67.89", CLASS_B_SUBNET));
	}

	@Test
	void matchesRange_class_C_subnet() {
		assertFalse(IpFilter.matchesRange("192.167.255.255", CLASS_C_SUBNET));
		assertTrue(IpFilter.matchesRange("192.168.0.0", CLASS_C_SUBNET));
		assertTrue(IpFilter.matchesRange("192.168.1.2", CLASS_C_SUBNET));
		assertTrue(IpFilter.matchesRange("192.168.255.255", CLASS_C_SUBNET));
		assertFalse(IpFilter.matchesRange("192.169.0.0", CLASS_C_SUBNET));
		assertFalse(IpFilter.matchesRange("123.45.67.89", CLASS_C_SUBNET));
	}

	@Test
	void matchesRange_all_IPv4() {
		assertTrue(IpFilter.matchesRange("0.0.0.0", ALL_SUBNET));
		assertTrue(IpFilter.matchesRange("1.2.3.4", ALL_SUBNET));
		assertTrue(IpFilter.matchesRange("255.255.255.255", ALL_SUBNET));
		assertTrue(IpFilter.matchesRange("123.45.67.89", ALL_SUBNET));
	}

	@Test
	void matchesRange_unexpected() {

		// ranges with unexpected subnet addresses (not aligned to subnet mask)
		assertTrue(IpFilter.matchesRange("127.0.0.0", "127.1.2.3/8"));
		assertTrue(IpFilter.matchesRange("10.0.0.0", "10.1.2.3/8"));
		assertTrue(IpFilter.matchesRange("172.16.0.0", "172.17.2.3/12"));
		assertTrue(IpFilter.matchesRange("192.168.0.0", "192.168.1.2/16"));
	}

	@Test
	void matchesRange_illegal_values() {
		assertFalse(IpFilter.matchesRange(null, "10.0.0.0/8"));
		assertFalse(IpFilter.matchesRange("10.0.0.1", null));
		assertFalse(IpFilter.matchesRange("10.0.0", "10.0.0.0/8"));
		assertFalse(IpFilter.matchesRange("10.0.0.1", "10.0.0/8"));
	}

	@Test
	void matchesAddress_IPv6() {
		assertTrue(IpFilter.matchesAddress("0123:4567:89ab:cdef:0123:4567:89ab:cdef", "0123:4567:89ab:cdef:0123:4567:89ab:cdef"));
		assertTrue(IpFilter.matchesAddress("fc00::1", "fc00::1"));
		assertTrue(IpFilter.matchesAddress("fc00::1", "FC00::1"));
		assertTrue(IpFilter.matchesAddress("fc00::1", "fc00:0:0:0:0:0:0:1"));
		assertTrue(IpFilter.matchesAddress("fc00::1", "fc00:0000:0000:0000:0000:0000:0000:0001"));
		assertTrue(IpFilter.matchesAddress("0:0:0:0:0:0:0:1", "::1")); // localhost
		assertFalse(IpFilter.matchesAddress("fc00::1", "fc00::2"));
	}

	@Test
	void matchesAddress_IPv6_mixed_with_IPv4() {
		assertFalse(IpFilter.matchesAddress("127.0.0.1", "::1"));
		assertFalse(IpFilter.matchesAddress("::1", "127.0.0.1"));
	}

	@Test
	void matchesAddress_IPv6_illegal_address() {
		assertFalse(IpFilter.matchesAddress("1::2::3", "1:2::3"));
		assertFalse(IpFilter.matchesAddress("1:2::3", "1::2::3"));
	}

	@Test
	void matchesRange_IPv6() {

		assertTrue(IpFilter.matchesRange("::", "::/0"));
		assertTrue(IpFilter.matchesRange("ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff", "::/0"));

		assertTrue(IpFilter.matchesRange("::", "::/128"));
		assertFalse(IpFilter.matchesRange("::1", "::/128"));

		assertFalse(IpFilter.matchesRange("::0", "::1/128"));
		assertTrue(IpFilter.matchesRange("::1", "::1/128"));
		assertFalse(IpFilter.matchesRange("::2", "::1/128"));

		assertFalse(IpFilter.matchesRange("::fffe:ffff:ffff:ffff", "::ffff:0:0:0/96"));
		assertTrue(IpFilter.matchesRange("::ffff:0:0:0", "::ffff:0:0:0/96"));
		assertTrue(IpFilter.matchesRange("::ffff:0:ffff:ffff", "::ffff:0:0:0/96"));
		assertFalse(IpFilter.matchesRange("::ffff:1:0:0", "::ffff:0:0:0/96"));

		assertFalse(IpFilter.matchesRange("64:ff9a:ffff:ffff:ffff:ffff:ffff:ffff", "64:ff9b::/96"));
		assertTrue(IpFilter.matchesRange("64:ff9b::", "64:ff9b::/96"));
		assertTrue(IpFilter.matchesRange("64:ff9b::ffff:ffff", "64:ff9b::/96"));
		assertFalse(IpFilter.matchesRange("64:ff9b::1:0:0", "64:ff9b::/96"));

		assertFalse(IpFilter.matchesRange("64:ff9b:0:ffff:ffff:ffff:ffff:ffff", "64:ff9b:1::/48"));
		assertTrue(IpFilter.matchesRange("64:ff9b:1::", "64:ff9b:1::/48"));
		assertTrue(IpFilter.matchesRange("64:ff9b:1:ffff:ffff:ffff:ffff:ffff", "64:ff9b:1::/48"));
		assertFalse(IpFilter.matchesRange("64:ff9b:2::0", "64:ff9b:1::/48"));

		assertFalse(IpFilter.matchesRange("ff:ffff:ffff:ffff:ffff:ffff:ffff:ffff", "100::/64"));
		assertTrue(IpFilter.matchesRange("100::", "100::/64"));
		assertTrue(IpFilter.matchesRange("100::ffff:ffff:ffff:ffff", "100::/64"));
		assertFalse(IpFilter.matchesRange("101::", "100::/64"));

		assertFalse(IpFilter.matchesRange("2000:ffff:ffff:ffff:ffff:ffff:ffff:ffff", "2001:0000::/32"));
		assertTrue(IpFilter.matchesRange("2001::", "2001:0000::/32"));
		assertTrue(IpFilter.matchesRange("2001::ffff:ffff:ffff:ffff:ffff:ffff", "2001:0000::/32"));
		assertFalse(IpFilter.matchesRange("2001:1::", "2001:0000::/32"));

		assertFalse(IpFilter.matchesRange("2001:1f:ffff:ffff:ffff:ffff:ffff:ffff", "2001:20::/28"));
		assertTrue(IpFilter.matchesRange("2001:20::", "2001:20::/28"));
		assertTrue(IpFilter.matchesRange("2001:2f:ffff:ffff:ffff:ffff:ffff:ffff", "2001:20::/28"));
		assertFalse(IpFilter.matchesRange("2001:30::", "2001:20::/28"));

		assertFalse(IpFilter.matchesRange("2001:db7:ffff:ffff:ffff:ffff:ffff:ffff", "2001:db8::/32"));
		assertTrue(IpFilter.matchesRange("2001:db8::", "2001:db8::/32"));
		assertTrue(IpFilter.matchesRange("2001:db8:ffff:ffff:ffff:ffff:ffff:ffff", "2001:db8::/32"));
		assertFalse(IpFilter.matchesRange("2001:db9::", "2001:db8::/32"));

		assertFalse(IpFilter.matchesRange("2001:ffff:ffff:ffff:ffff:ffff:ffff:ffff", "2002::/16"));
		assertTrue(IpFilter.matchesRange("2002::", "2002::/16"));
		assertTrue(IpFilter.matchesRange("2002:ffff:ffff:ffff:ffff:ffff:ffff:ffff", "2002::/16"));
		assertFalse(IpFilter.matchesRange("2003::", "2002::/16"));

		assertFalse(IpFilter.matchesRange("fbff:ffff:ffff:ffff:ffff:ffff:ffff:ffff", "fc00::/7"));
		assertTrue(IpFilter.matchesRange("fc00::", "fc00::/7"));
		assertTrue(IpFilter.matchesRange("fdff:ffff:ffff:ffff:ffff:ffff:ffff:ffff", "fc00::/7"));
		assertFalse(IpFilter.matchesRange("fe00::", "fc00::/7"));

		assertFalse(IpFilter.matchesRange("fe7f:ffff:ffff:ffff:ffff:ffff:ffff:ffff", "fe80::/64"));
		assertTrue(IpFilter.matchesRange("fe80::", "fe80::/64"));
		assertTrue(IpFilter.matchesRange("fe80::ffff:ffff:ffff:ffff", "fe80::/64"));
		assertFalse(IpFilter.matchesRange("fe80:0:0:1::", "fe80::/64"));

		assertFalse(IpFilter.matchesRange("feff:ffff:ffff:ffff:ffff:ffff:ffff:ffff", "ff00::/8"));
		assertTrue(IpFilter.matchesRange("ff00::", "ff00::/8"));
		assertTrue(IpFilter.matchesRange("ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff", "ff00::/8"));

	}

	@Test
	void matchesRange_IPv6_mixed_with_IPv4() {
		assertFalse(IpFilter.matchesRange("127.0.0.1", "::1/128"));
		assertFalse(IpFilter.matchesRange("::1", "127.0.0.1/8"));
	}

	@Test
	void matchesRange_IPv6_illegal_address() {
		assertFalse(IpFilter.matchesRange("1::2::3", "1:2::3/0"));
		assertFalse(IpFilter.matchesRange("1:2::3", "1::2::3/0"));
	}

}

src/test/java/net/markwalder/tomcat/AccessCheckTest.java
/*
 * MIT License
 *
 * Copyright (c) 2022 Stephan Markwalder
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package net.markwalder.tomcat;

import static org.junit.jupiter.api.Assertions.assertTrue;

import org.apache.catalina.connector.Request;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.junit.jupiter.MockitoExtension;

@ExtendWith(MockitoExtension.class)
class AccessCheckTest {

	@Mock
	Request request;

	AccessCheck accessCheck = new AccessCheck();

	@Test
	void test() {

		// test
		boolean result = accessCheck.test(request);

		// assert
		assertTrue(result);

		// verify
		Mockito.verifyNoInteractions(request);

	}

}

src/main/java/net/markwalder/tomcat/SessionLogoutListener.java
/*
 * MIT License
 *
 * Copyright (c) 2022 Stephan Markwalder
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package net.markwalder.tomcat;

import java.io.IOException;
import java.io.PrintWriter;
import java.security.Principal;
import java.util.Set;
import java.util.function.Function;
import java.util.function.Predicate;
import javax.servlet.ServletException;
import org.apache.catalina.Context;
import org.apache.catalina.Manager;
import org.apache.catalina.Session;
import org.apache.catalina.connector.Request;
import org.apache.catalina.connector.Response;
import org.apache.catalina.valves.ValveBase;
import org.apache.juli.logging.Log;
import org.apache.juli.logging.LogFactory;

public class SessionLogoutListener extends ValveBase {

	private final Predicate<Request> interceptor = new RequestInterceptor();
	private final Predicate<Request> accessCheck = (new RemoteAddrCheck(this::getIpFilter)).and(new PasswordCheck(this::getPassword));
	private final Function<Request, Set<String>> requestParser = new RequestParser();
	private final Log log;

	private String ipFilter = "127.0.0.1,::1";
	private String password = null;

	public SessionLogoutListener() {
		this(LogFactory.getLog(SessionLogoutListener.class));
	}

	// visible for testing
	SessionLogoutListener(Log log) {
		this.log = log;
	}

	public String getIpFilter() {
		return ipFilter;
	}

	@SuppressWarnings("unused") // used by Tomcat
	public void setIpFilter(String ipFilter) {
		this.ipFilter = ipFilter;
	}

	public String getPassword() {
		return password;
	}

	@SuppressWarnings("unused") // used by Tomcat
	public void setPassword(String password) {
		this.password = password;
	}

	@Override
	public void invoke(Request request, Response response) throws IOException, ServletException {

		// check if request is sent to session logout endpoint
		if (interceptor.test(request)) {

			handleRequest(request, response);
			return;
		}

		// forward request to next valve in the pipeline
		getNext().invoke(request, response);

	}

	private void handleRequest(Request request, Response response) throws IOException {

		// check if request is authenticated and authorized
		if (!accessCheck.test(request)) {
			sendResponse(403, "Forbidden", response);
			return;
		}

		// get usernames from request
		Set<String> usernames = requestParser.apply(request);
		if (!usernames.isEmpty()) {

			// logout all users with the given usernames
			Context context = request.getContext();
			logoutUsers(context, usernames);

		}

		// return OK message and stop request processing
		sendResponse(200, "OK", response);
	}

	private void logoutUsers(Context context, Set<String> usernames) {

		if (log.isDebugEnabled()) {
			log.debug("usernames: '" + String.join("', '", usernames) + "'");
		}

		// get all Tomcat sessions for the current webapp context
		Session[] sessions = getAllSessions(context);

		// for every session ...
		for (Session session : sessions) {

			// ignore sessions that have already been invalidated
			if (!session.isValid()) {
				continue;
			}

			// ignore sessions for unauthenticated users
			Principal principal = session.getPrincipal();
			if (principal == null) {
				continue;
			}

			// ignore sessions for other users
			String principalName = principal.getName();
			if (!usernames.contains(principalName)) {
				continue;
			}

			// remember session ID
			String sessionId = session.getId();

			// logout the session
			session.expire();

			if (log.isDebugEnabled()) {
				String truncatedSessionId = truncateSessionId(sessionId); // log only first 8 characters of session ID
				log.debug("session: id='" + truncatedSessionId + "...', principal='" + principalName + "'");
			}

		}
	}

	private static Session[] getAllSessions(Context context) {
		Manager manager = context.getManager();
		return manager.findSessions();
	}

	public static String truncateSessionId(String sessionId) {
		return sessionId.substring(0, 8);
	}

	private static void sendResponse(int status, String message, Response response) throws IOException {
		response.setStatus(status);
		response.setContentType("text/plain");
		response.setCharacterEncoding("UTF-8");
		PrintWriter writer = response.getWriter();
		writer.print(message);
	}

}
